### ISSUE TITLE
Obtaining the dependents of a given package

### ISSUE THREAD

**Issue by pathob** [conan-io/conan#3475]
Hi,

I'm currently figuring out how to work with conan and need to obtain the dependents of a given package. I've read your documentation ([LINK](https://docs.conan.io/en/latest/faq/using.html#how-to-obtain-the-dependents-of-a-given-package)) about that topic and understood that the dependency graph can be different for every configuration. Nevertheless, if I understood it right, the dependency graph should be obtainable for a given package ID.

But how can I pass the package ID to the `conan info` command? The format mentioned in the documentation is `MyLib/0.1@user/channel:63da998e3642b50bee33`. When I'm trying to do `conan info MyLib/0.1@user/channel:63da998e3642b50bee33`, this results in the following error:

```
ERROR: Conanfile not found: /path/to/MyLib/MyLib/0.1@user/channel:63da998e3642b50bee33
```

I know the package ID is too short, I substituted the real values to match the documentation.

In a first step, I would really be happy if you would tell me how to do that correctly. And in a second step, the documentation should be updated in order to help other users in the future.

Thank you!

**Environment**
Conan: 1.7.1
OS: Fedora 28 (installed via PIP)

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by pathob** [comment]
Hi @memsharded 

thanks for the fast response and the good example. But I still don't manage to obtain the dependents. Can I somehow pass the settings from a conaninfo.txt to `conan info / install`?

**Comment by memsharded** [comment]
Yes, if you ran a ``conan install`` first, it created the ``conaninfo.txt`` somewhere. Pass it to the command:

```bash
conan info . --install-folder=/path/to/folder
```

If you want, you can also provide the same input profile or settings, as you did with ``conan install``:
```bash
conan info . --profile=myprofile # or -s setting=value ...
```

**Comment by pathob** [comment]
When I don't provide any settings or a profile, still the default settings are used, right?

I'm executing `conan info .` in the directory (and install directory) of a component but I'm only getting

```
Component1/master@PROJECT
    ID: 2d5685a99e2ff5a0d2b054bb1fbfa46781f8a46a
    BuildID: None
    URL: https://github.com/testcca/Component1.git
    License: <Put the package license here>
```

Executing `conan info .` on the project (with the same default settings) that is using that component gives me:

```
Project/master@PROJECT
    ID: 7baaa60590b1aa6cf070ec535a154cc8af509ad1
    BuildID: None
    URL: https://github.com/testcca/Project.git
    License: MIT
    Requires:
        Component1/master@conti/stable
Component1/master@conti/stable
    ID: 2d5685a99e2ff5a0d2b054bb1fbfa46781f8a46a
    BuildID: None
    Remote: artifactory=http://localhost:8081/artifactory/api/conan/conan-local
    URL: https://github.com/testcca/Component1.git
    License: <Put the package license here>
    Recipe: Cache
    Binary: Cache
    Binary remote: artifactory
    Creation date: 2018-09-04 14:55:46
    Required by:
        Project/master@PROJECT
```

What can I do to see all projects that are using the component?

Maybe there is something wrong in my test setup that can be found at https://github.com/testcca

**Comment by memsharded** [comment]
Yes, if you don't provide any settings or options, the default profile is being used. Actually, when you provide some settings or options, they are overriding the values from the default profile.

It is not possible to know the dependants of a given package either. There is no direct downstream mapping for packages dependants. You would need to processs and store all possible configurations from all the packages in the repo to have some idea about it, as there are conditional requirements; you can depend on another package based on your settings or options, for example. And that would only apply to a given repo, but conan is decentralized, it is usual to have packages in different repos.

**Comment by pathob** [comment]
I understand what you say about the settings and the fact that conan is decentralized. But still, I think conan **could** be able to give me the desired information.

Say we are using only a single repository and are processing and uploading all packages by ourselfs. Then we would have all the required information, right?

Our use case is to rebuild all projects that are dependent on a component (library) that has changed.

**Comment by memsharded** [comment]
Let me give an example:

```python
class PkgA(ConanFile):
    pass
```
```python
class PkgB(ConanFile):
    build_requires = "testlib/0.1@user/stable"
   def requirements(self):
       if self.settings.os == "Windows" and self.settings.compiler == "gcc":
            self.requires("PkgA/0.1@user/channel")
      elif self.options.shared:
           self.requires("PkgA/1.2@user/channel")
      #Note that otherwise there will be no requirement to PkgA
```
Now, the question: Is PkgB a dependant of PkgA? What should the system give as an answer to "give me the dependants on "PkgA/0.1"? Or "PkgA/1.2"? What are the dependants on "testlib/0.1"? Because a build_require is actually not necessary, so if there are already binaries for PkgB, then PkgB is not a dependant on "testlib".

Now propagate that downstream with all the combinatorics that can happen. 

So the way to approach this issue is to have defined "consumer" packages or applications. Also, because it is the only way to manage the complexity for multi-project. Lets say that you have a library, that is shared between 2 or more projects. It is very likely that you want to do some development on it, and have your CI fire on a given project, but not the whole repo projects (including old projects!). Then, when things look good in that project, you might considering doing some further release of that package to be used in the other projects too.

So the questions that are approachable are of the kind: for this "project" (a conanfile), what are the dependencies that need to be rebuilt, given that "PkgA" has had some modifications, for a given configuration. This is what can be done with something like (check reference here: https://docs.conan.io/en/latest/reference/commands/consumer/info.html):

```bash
$ conan info myproject.py --build-order=PkgA
```

The interface to this command is something that will be improved. But still, the way to get this information will be from downstream "consumers", it is not possible to visit the graph from upstream to downstream due to such conditionally.

**Comment by memsharded** [comment]
Long time has passed, we learned a lot in the meantime:

- Processing dependents of packages is something that is really not feasible. Modern processes and devops is quite decentralized, it uses multiple repositories and multiple servers in many cases
- The combinatoric explosion of conditions, versions and revisions downstream makes any possible solution not scale. You modify something like zlib/1.2.X and it will tell you that it has tens of thousands of dependants

The process for doing CI at scale has settled on the principle that it is necessary to define the consumer "products", that is several packages, versions and configurations to support. Given those "products" Conan has now tools to evaluate what needs to be built, to aggregate builds to be optimal, and to distribute the build in a safe and robust way with lockfiles. Some initial docs can be seen here: https://docs.conan.io/2/tutorial/versioning.html, and more docs for the specific CI case will be added in the following weeks/months.

Closing this issue as won't implement, asking servers about dependents will not be possible.
